### DISCUSSION TITLE
Form data along with FileUpload not working in FastAPI

### DISCUSSION THREAD

**Discussion by sandchak** [fastapi/fastapi#12525] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile, Form, status, Response, File, Depends
from fastapi.responses import RedirectResponse, HTMLResponse
from pydantic import BaseModel
from enum import Enum
from typing import Annotated


app = FastAPI()

class Sex(Enum):
    Male = "Male"
    Female = "Female"

class FormData(BaseModel):
    name: str
    age: int
    sex: Sex
    address: str

@app.post("/save", status_code=status.HTTP_201_CREATED)
def post(data: Annotated[FormData, Form()], image: UploadFile = File(...)) -> Response:
    if data:
        print(data.model_dump(mode='json'))
        print(image.filename)
        return {"user": data.name, "status": "Created"}
    
    return RedirectResponse("http://localhost:8000/")


@app.get("/")
def main():
    content = '''
<body>
<form action="/save" enctype="multipart/form-data" method="post">
<input name="name" type="text" />
<input name="age" type="number" min=18 step=1 />
<select name="sex">
  <option value="Male">Male</option>
  <option value="Female">Female</option>
</select>
<input name="address" type="text" />
<input name="image" type="file" />
<input type="submit" />
</form>
</body>
'''
    return HTMLResponse(content=content)
```

### Description

I am trying to receive Form data along with File data in my FastAPI code, but its not working.... Kindly suggest...
I getting below error while trying it from browser...
{
  "detail": [
    {
      "type": "missing",
      "loc": [
        "body",
        "data"
      ],
      "msg": "Field required",
      "input": null
    }
  ]
}

### Operating System

Windows

### FastAPI Version

0.115.2

### Pydantic Version

2.9.2

### Python Version

Python 3.10.11

**Comment by 1001pepi** [comment]
For Body parameters, fast API can infer types and understand how to restructure data provided in the request to meet the format defined by the pydantic BaseModel derived class by reading the data provided in the request on the JSON format. 

Since your HTML form uses the multipart/form-data enctype, the data provided in the request is not on a JSON format, that is why the mappy to the BaseModel derived class is not working. 

Instead you can try to get each field of the form separately inn the function like this:

```python
from fastapi import FastAPI, UploadFile, Form, status, Response, File
from fastapi.responses import HTMLResponse
from typing import Annotated


app = FastAPI()

@app.post("/save", status_code=status.HTTP_201_CREATED)
def post(
    name: Annotated[str, Form()], 
    age: Annotated[int, Form()],
    sex: Annotated[str, Form()],
    address: Annotated[str, Form()],
    image: UploadFile = File(...)) -> Response:
    
    print(f"name: {name}, age: {age}, sex: {sex}, address: {address}")
    print(f"Image name: {image.filename}")
    
    return "OK"


@app.get("/")
def main():
    content = '''
<body>
<form action="/save" enctype="multipart/form-data" method="post">
<input name="name" type="text" />
<input name="age" type="number" min=18 step=1 />
<select name="sex">
  <option value="Male">Male</option>
  <option value="Female">Female</option>
</select>
<input name="address" type="text" />
<input name="image" type="file" />
<input type="submit" />
</form>
</body>
'''
    return HTMLResponse(content=content)
```

**Comment by LincolnPuzey** [comment]
Its not documented, but it seems that right now using a [Form Model](https://fastapi.tiangolo.com/tutorial/request-form-models/) and `UploadFile`(s) as independent path operation function parameters is not supported.

I think https://github.com/fastapi/fastapi/discussions/12138 is the same issue as you are having. There are suggestions there for how to make it work.

**Comment by gbaian10** [comment]
I expected it to work when combined with the BaseModel for the same issue, but it didn't.
At least, the documentation gave me the impression that it was possible.
